Add `CUDA_KERNEL_ASSERT_PRINTF`, a more flexible  `CUDA_KERNEL_ASSERT_MSG`

mjkatmeta opened the pull request:
This new assertion helper bundles a printf call with the assertion. The goal is to make changes to instrument asserts with device-side information more intuitive and less error-prone. (See the printf call in ATen/native/cuda/Repeat.cu.) Parametrized error messages are a substantial improvement in debuggability because they show the mismatched device-side values. This lets us avoid a whole cycle of rebuilding + re-running failing training workflows.

We include file, line number, function, and failing condition in the printf (along with the message provided by the user). The format matches the format of the message output by `__assert_fail`. There's also an easy-to-grep-for keyword `CUDA_KERNEL_ASSERT` in the message.

I'm following the existing patterns of arch-specific macros - e.g., on ROCm, this is just a call to abort(), just like the other `CUDA_KERNEL_ASSERT*` variations. I'd appreciate any thoughts on architecture-specific testing (most likely on the OSS side).

# Alternatives
* We could just update `CUDA_KERNEL_ASSERT_MSG`. That would mean introducing `printf` calls from the kernel where there weren't any before, though. This seems like a bad idea because of the performance sensitivity.
* If we want to move more slowly here, I could instrument more `CUDA_KERNEL_ASSERT` callsites without a macro, similar to https://github.com/pytorch/pytorch/pull/157996. But the main downside here is the performance hit, so let's have an organized way of doing it first.

# Risks/Problems
* We're shoving a lot of stuff into this printf. If a filename (at compile-time) contains `%s`, we will end up dereferencing whatever value was pushed in. On a CPU this can cause a segfault. I don't know how it behaves on a GPU.
* Adding printf calls can have a performance impact because of increased register and stack usage. I did not see this play out in practice (see "benchmarks" below). However, there are changes to the generated PTX that could result in performance problems later (see "changes in generated PTX" below).

# Benchmarks

* I ran the following benchmarks a several times on a host with an A100: https://gist.github.com/mjkatmeta/e5494d949204a2afe2d43c452b99424f
* Results are here -- I couldn't find a significant difference before or after  https://gist.github.com/mjkatmeta/0f99ec27bb91214fb2cc7f612938d431

# Change in generated PTX

This is the easiest way I found to run nvcc over just Repeat.cu (this is a buck2 target that includes just a copy of Repeat.cu):
```
buck2 build --show-output scripts/mjk/ai_training/cuda_benchmarks:repeat_cuda
# then use the printed .so file like this:
~/fbsource/third-party/cuda/cuda_12.8.0/x64-linux/bin/cuobjdump -ptx ../buck-out/v2/gen/fbcode/028bde1acfaba823/scripts/mjk/ai_training/cuda_benchmarks/__repeat_cuda__/libscripts_mjk_ai_training_cuda_benchmarks_repeat_cuda.so
```

## with printf
This is the version of the code that appears in this diff:

https://gist.github.com/mjkatmeta/5d18d48282d46b2240d946b335052b9a

## without printf
I recompiled, replacing `CUDA_KERNEL_ASSERT_PRINTF(...)` in Repeat.cu with:
```
CUDA_KERNEL_ASSERT(result_size == cumsum_ptr[size - 1]);
```
https://gist.github.com/mjkatmeta/480df4b3a122e7b326554dd15ebb7c9d

(Both of these are annotated with `// CHAR ARRAY:` comments to make the string constants easier to read.)

Test Plan:
Running this minimal test case:

```
import torch
def main():
    x = torch.ones(10, dtype=torch.int64, device="cuda:0")
    torch.repeat_interleave(x, x, output_size=0)
```

Now we see the new message (from printf) alongside the assert failure:

```
$ buck2 run fbcode//scripts/darshanr/repeat_interleave_errors:repeat_interleave_errors
[...]
[CUDA_KERNEL_ASSERT] fbcode/caffe2/aten/src/ATen/native/cuda/Repeat.cu:25: compute_cuda_kernel: block: [0,0,0], thread: [31,0,0]: Assertion failed: `result_size == cumsum_ptr[size - 1]`: Invalid input! In `repeat_interleave`, the `output_size` argument (0) must be the same as the sum of the elements in the `repeats` tensor (10).

fbcode/caffe2/aten/src/ATen/native/cuda/Repeat.cu:25: compute_cuda_kernel: block: [0,0,0], thread: [384,0,0] Assertion `result_size == cumsum_ptr[size - 1]` failed.
[...[
```

Rollback Plan:

Reviewed By: mradmila

Differential Revision: D79310684

mjkatmeta commented:
I'll look at the lint errors.

janeyx99 commented:
Using printf causes a perf regression during runtime because, as I understand it, its arguments need to be evaluated at runtime even if printf is never triggered, regardless of architecture. Therefore, it is not good to introduce printfs everywhere in the codebase. The tradeoff with perf vs a helper macro may not be worth it.

mjkatmeta commented:
> Using printf causes a perf regression during runtime because, as I understand it, its arguments need to be evaluated at runtime even if printf is never triggered, regardless of architecture. Therefore, it is not good to introduce printfs everywhere in the codebase. The tradeoff with perf vs a helper macro may not be worth it.

OK, I was worried that might be the case. I'd like to learn more about this - I'll ping you separately. Thanks!

janeyx99 commented:
cc @ngimel @eqy @syed-ahmed who would know about the impact of printf on CUDA kernels more than I would.

mjkatmeta commented on torch/headeronly/macros/Macros.h:
```diff
@@ -395,6 +396,18 @@ __host__ __device__
   }
+#define CUDA_KERNEL_ASSERT_PRINTF(cond, msg, ...)                           \
+  if (C10_UNLIKELY(!(cond))) {                                              \
+    (void)(printf(                                                          \
```
My understanding here is pretty shallow so I double-checked this -- yes, fmt::printf appears to be host-only.

I tried importing `fmt/printf.h` and using `fmt::printf`, I get the following error:
```
    fbcode/caffe2/aten/src/ATen/native/cuda/Repeat.cu(20): error: calling a __host__ function("int fmt::v9::printf<char [319], unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, long, long , (int)0> (const T1 &, const T2 &...)") from a __global__ function("compute_cuda_kernel<long> ") is not allowed
```

mjkatmeta commented on aten/src/ATen/native/cuda/Repeat.cu:
```diff
@@ -17,12 +17,18 @@ __global__ static void compute_cuda_kernel(
+    threadIdx.z,
+    result_size,
+    cumsum_ptr[size - 1 ]
+  );
```
To add a little background: I picked this location because this is where @drdarshan instrumented the kernel with a printf manually. This extra output helped debugging a production training issues that was ultimately caused by bad data (the exact sizes of the values being compared were useful to folks who knew the model).

mjkatmeta commented:
Update this to include `blockId` and `threadId`. From what I've seen, dumping these out is pretty standard practice when manually adding printfs to aid in debugging. These should always be available in a kernel, and you should get a compile-time error if not.

ngimel commented:
Yeah indexing kernel would be a good candidate for examination, creating some reasonably-sized indexing test case so that kernel runs for 20-40us and benchmarking it and examining in ncu would be good.

eqy commented:
I consulted with some NV folks and the feedback is that the `printf`, even if never executed would be expected to negatively impact perf. due to increasing register usage, printf-related housekeeping, dependencies on registers, and serialization with other `printfs`. One suggestion was to always `#ifdef` the `printf`s.

mjkatmeta commented:
After a discussion with @drdarshan, we decided that (3) is a really bad option (special `#defines` for FB-internal use cases). I'm going to continue with the benchmark option and see how much of a regression we see.

I'll also examine the generated ptx and see if we can interpret the diffs.

ngimel commented:
Regressions might be different between kernels that are already register-heavy and kernels that don't use too many registers. It's hard to benchmark it conclusively, and given that those asserts typically don't provide much info outside "your data was wrong", I'd much rather have it as an option than constantly think if it regresses perf or not

mjkatmeta commented:
I understand the hesitation to introduce an easy way to add device-side printfs if the performance impact is unknown.

Here's where I am and what I'm looking at next:

1. I have a very basic benchmark that runs the repeat_interleave kernel (simply because it was very easy for me to map pytorch code to the kernel). But now that I understand your point about the register pressure effect on different kernels being different, I'm not sure that's useful.
2. I'm looking at generated PTX code with and without added printfs, but it's not clear to me that it'll be easy to accurately judge the register pressure changes from the diffs. We definitely have more `.reg` calls for params. I'm still learning how to read PTX, though, so this is going slowly.
3. I'll try to run `nvcc` with `--resource-usage` over the kernels with and without the printf calls. I'm hoping this tells me something useful.

If none of this turns up anything interesting, `#ifdef`-ing it away sounds like the only remaining option.

To explain why I keep pressing on this: From my (admittedly limited) experience debugging large incidents caused by corrupted data, getting more info from assertion failures up-front can help a lot. For example, it helps us understand whether a new batch of failing training runs is due to a known problem, or another, new data problem. Being able to see added debugging information *without* re-launching every failing workflow with a test version (either with manually-added debugging, or to flip a `#define`) has a substantial benefit in these cases.

mjkatmeta commented:
I re-ran benchmarks using what I believe is a more trustworthy method, and against more kernels -- I'm using  `torch.utils.benchmark.Timer` now. I don't see a performance impact anymore:

* I ran the following benchmarks a several times on a host with an A100: https://gist.github.com/mjkatmeta/e5494d949204a2afe2d43c452b99424f
* Results are here: https://gist.github.com/mjkatmeta/0f99ec27bb91214fb2cc7f612938d431

I am not sure how to resolve the conflicting opinions on how we should proceed:
1. We have agreement from AI Infra folks (@drdarshan, @mradmila) that this is worth trying out. @drdarshan has concerns about the maintainability of using the `#ifdef`.
2. @ngimel and @eqy raised concerns about performance of the prinfs.

Are there different performance tests I can run that would definitively answer the performance question? What is the impact if we modify one of the kernels and then, say, we see a real regression? I hope we collectively agree on a path forward. Thanks!

mjkatmeta committed:
Add `CUDA_KERNEL_ASSERT_PRINTF`, a more flexible  `CUDA_KERNEL_ASSERT_MSG` (#160129)

Summary:

This new assertion helper bundles a printf call with the assertion. The goal is to make changes to instrument asserts with device-side information more intuitive and less error-prone. (See the printf call in ATen/native/cuda/Repeat.cu.) Parametrized error messages are a substantial improvement in debuggability because they show the mismatched device-side values. This lets us avoid a whole cycle of rebuilding + re-running failing training workflows.

We include file, line number, function, and failing condition in the printf (along with the message provided by the user). The format matches the format of the message output by `__assert_fail`. There's also an easy-to-grep-for keyword `CUDA_KERNEL_ASSERT` in the message.

I'm following the existing patterns of arch-specific macros - e.g., on ROCm, this is just a call to abort(), just like the other `CUDA_KERNEL_ASSERT*` variations. I'd appreciate any thoughts on architecture-specific testing (most likely on the OSS side).

# Alternatives
* We could just update `CUDA_KERNEL_ASSERT_MSG`. That would mean introducing `printf` calls from the kernel where there weren't any before, though. This seems like a bad idea because of the performance sensitivity.
* If we want to move more slowly here, I could instrument more `CUDA_KERNEL_ASSERT` callsites without a macro, similar to https://github.com/pytorch/pytorch/pull/157996. But the main downside here is the performance hit, so let's have an organized way of doing it first.

# Risks/Problems
* We're shoving a lot of stuff into this printf. If a filename (at compile-time) contains `%s`, we will end up dereferencing whatever value was pushed in. On a CPU this can cause a segfault. I don't know how it behaves on a GPU.
* Adding printf calls can have a performance impact because of increased register and stack usage. I did not see this play out in practice (see "benchmarks" below). However, there are changes to the generated PTX that could result in performance problems later (see "changes in generated PTX" below).

# Benchmarks

* I ran the following benchmarks a several times on a host with an A100: https://gist.github.com/mjkatmeta/e5494d949204a2afe2d43c452b99424f
* Results are here -- I couldn't find a significant difference before or after  https://gist.github.com/mjkatmeta/0f99ec27bb91214fb2cc7f612938d431

# Change in generated PTX

This is the easiest way I found to run nvcc over just Repeat.cu (this is a buck2 target that includes just a copy of Repeat.cu):
```
buck2 build --show-output scripts/mjk/ai_training/cuda_benchmarks:repeat_cuda
# then use the printed .so file like this:
~/fbsource/third-party/cuda/cuda_12.8.0/x64-linux/bin/cuobjdump -ptx ../buck-out/v2/gen/fbcode/028bde1acfaba823/scripts/mjk/ai_training/cuda_benchmarks/__repeat_cuda__/libscripts_mjk_ai_training_cuda_benchmarks_repeat_cuda.so
```

## with printf
This is the version of the code that appears in this diff:

https://gist.github.com/mjkatmeta/5d18d48282d46b2240d946b335052b9a

## without printf
I recompiled, replacing `CUDA_KERNEL_ASSERT_PRINTF(...)` in Repeat.cu with:
```
CUDA_KERNEL_ASSERT(result_size == cumsum_ptr[size - 1]);
```
https://gist.github.com/mjkatmeta/480df4b3a122e7b326554dd15ebb7c9d

(Both of these are annotated with `// CHAR ARRAY:` comments to make the string constants easier to read.)

Test Plan:
Running this minimal test case:

```
import torch
def main():
    x = torch.ones(10, dtype=torch.int64, device="cuda:0")
    torch.repeat_interleave(x, x, output_size=0)
```

Now we see the new message (from printf) alongside the assert failure:

```
$ buck2 run fbcode//scripts/darshanr/repeat_interleave_errors:repeat_interleave_errors
[...]
[CUDA_KERNEL_ASSERT] fbcode/caffe2/aten/src/ATen/native/cuda/Repeat.cu:25: compute_cuda_kernel: block: [0,0,0], thread: [31,0,0]: Assertion failed: `result_size == cumsum_ptr[size - 1]`: Invalid input! In `repeat_interleave`, the `output_size` argument (0) must be the same as the sum of the elements in the `repeats` tensor (10).

fbcode/caffe2/aten/src/ATen/native/cuda/Repeat.cu:25: compute_cuda_kernel: block: [0,0,0], thread: [384,0,0] Assertion `result_size == cumsum_ptr[size - 1]` failed.
[...[
```

Rollback Plan:

Reviewed By: cnphil, mradmila

Differential Revision: D79310684
